synapse/storage/databases/main/search.py

```diff
@@ -770,20 +790,168 @@ def _to_postgres_options(options_dict: JsonDict) -> str:
     return "'%s'" % (",".join("%s=%s" % (k, v) for k, v in options_dict.items()),)
 
 
-def _parse_query(database_engine: BaseDatabaseEngine, search_term: str) -> str:
-    """Takes a plain unicode string from the user and converts it into a form
-    that can be passed to database.
-    We use this so that we can add prefix matching, which isn't something
-    that is supported by default.
+@dataclass
+class Phrase:
+    phrase: List[str]
+
+
+class Not:
+    pass
+
+
+class Or:
+    pass
+
+
+class And:
+    pass
+
+
+class LParen:
+    pass
+
+
+class RParen:
+    pass
+
+
+Token = Union[str, Phrase, Not, Or, And, LParen, RParen]
+TokenList = List[Token]
+
+
+def _tokenize_query(query: str) -> TokenList:
     """
+    Convert the user-supplied `query` into a TokenList, which can be translated into
+    some DB-specific syntax.
 
-    # Pull out the individual words, discarding any non-word characters.
-    results = re.findall(r"([\w\-]+)", search_term, re.UNICODE)
+    The following constructs are supported:
 
-    if isinstance(database_engine, PostgresEngine):
-        return " & ".join(result + ":*" for result in results)
-    elif isinstance(database_engine, Sqlite3Engine):
-        return " & ".join(result + "*" for result in results)
+    - phrase queries using "double quotes"
+    - case-insensitive or and and operators, with parentheses to change operator precedence
+    - negation of a keyword via unary -
+    - unary hyphen to denote NOT e.g. 'include -exclude'
+
+    """
+    tokens: TokenList = []
+    words = query.split(" ")
+    i = 0
+    while i < len(words):
+        word = words[i]
+        if word[0] == '"':
+            phrase_word = word[1:]
+            phrase = []
+            while True:
+                if phrase_word[-1] == '"':
+                    phrase.append(phrase_word[:-1])
+                    break
+                else:
+                    phrase.append(phrase_word)
+                i = i + 1
+                if i == len(words):
+                    break
+                phrase_word = words[i]
+            tokens.append(Phrase(phrase))
+        elif word[0] == "-":
+            tokens.append(Not())
+            tokens.append(word[1:])
+        elif word.lower() == "or":
+            tokens.append(Or())
+        elif word.lower() == "and":
+            tokens.append(And())
+        elif word == "(":
```

Comment:
Should this not check just the start of the word? So that it matches `(foo AND bar)`?

Reply:
I think this is getting lucky that the syntax is essentially the same for websearch vs. SQL search here, so we fall-through to the "last" case (which just appends `"(foo"`, then when we convert back to SQL syntax append that token back out so you get `(foo | bar)` instead of `( foo | bar )`.

This will breakdown though if you do something like `("f o o" and bar)` though I *think*?